documentation/Architecture/The-Data-Processing-Pipeline.md

```diff
@@ -1,43 +1,131 @@
+
+<img src=".attachments/cooker_query.svg" width="800">
+
+In this example, first the `DataCooker` declares a `DataOutput` of type `int` whose name is `Foo`. Then, the `Table` 
+asks the SDK for that `DataOutput`. As a result, the `Table` receives `Foo`.
+
+To query a `DataOutput`, the querying code only needs to know
+1) The `DataCookerPath` for the `DataCooker` it wishes to receive data from
+2) The (`string`) name of the `DataOutput` property of that `DataCooker`
+3) The type of the `DataOutput` property
+
+Because the querying code only needs these three properties, external assemblies - even those without access to 
+any `DataCooker`s' source code - can query the SDK for data from any `DataCooker` which is loaded.
+
+Finally, the SDK facilitates both the transfer of data and ensuring that a queried `DataCooker`'s `DataOutput` is ready. If, for 
+example, the `DataCooker` above receives input from a `SourceParser`, the SDK will ensure that the `SourceParser` has processed 
+all of its data sources - and that the `DataCooker` has received ("cooked") all of its required data - *before* the `DataOutput` 
+is sent to the `Table`.
+
+
+### Data Cooker Types
+
+As mentioned above, a `DataCooker` can consume data from either a single `SourceParser` or one or more other `DataCooker`s. 
+The SDK handles these cases differently, so it is important to understand the differences between them.
+
+#### Source Data Cookers
+
+A `SourceDataCooker` is a `DataCooker` which consumes data from a single `SourceParser`. It must specify
+1) The `SourceParser` to consume data from and
+2) The keys of the events it wishes to receive
+
+It is helpful to think of keys as "group names" here. A `SourceDataCooker` specifies the "groups," or "types," of 
+events it wishes to receive. 
+
+Finally, a `SourceDataCooker` will consume (cook) events *one at a time* as the `SourceParser` emits them
+
+#### Composite Data Cookers
+
+A `CompositeDataCooker` is a `DataCooker` which consumes data from one or more other `DataCooker`s. It must 
+specify all of the other `DataCookerPath`s for the `DataCooker`s which it depends upon. The SDK instructs
+the `CompositeDataCooker` when *all* of its dependencies have finished processing their data, at which point 
+the `CompositeDataCooker` can query the `DataOutput`s it needs to perform *its* transformations.
+
+# Putting Everything Together
+
+By combining `SourceParser`s, `SourceDataCooker`s, and `CompositeDataCooker`s, a plugin can create arbitrarily 
+complex and extensible data processing pipelines. For example, here is a pipeline a plugin may create to 
+modularize and promote the extensibility of three tables:
+
+<img src=".attachments/complex_pipeline.svg" width="100%">
+
+In this example, `DC1` and `DC2` are `SourceDataCooker`s which are concurrently consuming events emitted from the 
+`SourceParser`. `T1` queries `DC1`'s `DataOutput1` property to construct its table. `DC3` and `DC4` are 
```

Comment:
clarify T1, is Table 1 more explicitly at the beginning of this explanation or instead of using T1.

Reply:
Good catch!